xarray/core/datatree.py

```diff
@@ -459,7 +471,9 @@ def __init__(
             children = {}
 
         super().__init__(name=name)
-        self._set_node_data(_coerce_to_dataset(data))
+        ds = _coerce_to_dataset(data)
+        self._set_node_data(ds)
+        _check_for_slashes_in_names(ds.variables)
```

Comment:
Perhaps the `_check_for_slashes_in_names` should just live inside the `._set_node_data` method, if that is the only place it would ever be used.